File: module-quote.1/MyQuote/app/src/main/java/com/developer/cookie/myquote/quote/fragments/CurrentQuoteFragment.java
package com.developer.cookie.myquote.quote.fragments;


import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.ShareActionProvider;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.developer.cookie.myquote.R;
import com.developer.cookie.myquote.database.QuoteDataRepository;
import com.developer.cookie.myquote.database.model.QuoteText;
import com.developer.cookie.myquote.quote.Types;
import com.developer.cookie.myquote.utils.FillViewsWithCurrentQuoteDataHelper;

import io.realm.RealmChangeListener;
import io.realm.RealmResults;

/**
 * CurrentQuoteFragment is used for display information about quote.
 */
public class CurrentQuoteFragment extends Fragment {
    private static final String LOG_TAG = CurrentQuoteFragment.class.getSimpleName();
    private static final String CURRENT_QUOTE_ID_NEW_INSTANCE = "current_quote_fragment_current_quote_id_new_instance";
    private static final String CURRENT_QUOTE_TYPE_NEW_INSTANCE = "current_quote_fragment_current_quote_type_new_instance";
    private static final String QUOTE_TYPE_BUNDLE = "current_quote_fragment_quote_type_bundle";
    private static final String QUOTE_ID_BUNDLE = "current_quote_fragment_quote_id_bundle";
    Long mCurrentQuoteId;
    QuoteDataRepository mQuoteDataRepository;
    RealmResults<QuoteText> mCurrentQuote;
    String mCurrentCategory;
    String mQuoteType;
    ShareActionProvider mShareActionProvider;
    Intent sharingIntent;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        defineInputData(savedInstanceState);
        mQuoteDataRepository = new QuoteDataRepository();
        mCurrentQuote = mQuoteDataRepository.getQuoteByQuoteId(mCurrentQuoteId);
    }

    private void defineInputData(Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            mQuoteType = savedInstanceState.getString(QUOTE_TYPE_BUNDLE);
            mCurrentQuoteId = savedInstanceState.getLong(QUOTE_ID_BUNDLE);
        } else if (getArguments() != null) {
            mQuoteType = getArguments().getString(CURRENT_QUOTE_TYPE_NEW_INSTANCE);
            mCurrentQuoteId = getArguments().getLong(CURRENT_QUOTE_ID_NEW_INSTANCE);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View rootView = inflater.inflate(R.layout.fragment_current_quote, container, false);
        final TextView quoteTextView = (TextView) rootView.findViewById(R.id.current_quote_text);
        final TextView bookNameView = (TextView) rootView.findViewById(R.id.current_book_name);
        final TextView authorNameView = (TextView) rootView.findViewById(R.id.current_author_name);
        final TextView pageNumberView = (TextView) rootView.findViewById(R.id.current_page_number);
        final TextView publisherNameTextView = (TextView) rootView.findViewById(R.id.current_publisher_name);
        final TextView yearNumberView = (TextView) rootView.findViewById(R.id.current_year_number);
        final TextView quoteCreationDate = (TextView) rootView.findViewById(R.id.quote_creation_date);
        final TextView currentCategory = (TextView) rootView.findViewById(R.id.current_category);
        mCurrentQuote.addChangeListener(new RealmChangeListener<RealmResults<QuoteText>>() {
            @Override
            public void onChange(RealmResults<QuoteText> element) {
                if (element.size() > 0) {
                    FillViewsWithCurrentQuoteDataHelper.fillViewsWithCurrentQuoteData(element, quoteTextView,
                            bookNameView, authorNameView, pageNumberView, publisherNameTextView, yearNumberView, mQuoteType);
                    setQuoteCategory(currentCategory, element);
                    setQuoteCreationDate(quoteCreationDate, element);
                    hideFieldForMyQuotes(rootView);
                }
            }
        });
        return rootView;
    }

    private void setQuoteCategory(TextView currentCategory, RealmResults<QuoteText> element) {
        mCurrentCategory = element.first().getCategory().getCategoryName();
        currentCategory.setText(element.first().getCategory().getCategoryName());
        currentCategory.setAllCaps(true);
    }

    private void setQuoteCreationDate(TextView quoteCreationDate, RealmResults<QuoteText> element) {
        quoteCreationDate.setText(element.first().getDate().getQuoteDateValue());
    }

    private void hideFieldForMyQuotes(View rootView) {
        if (mQuoteType.equals(Types.MY_QUOTE)) {
            LinearLayout quoteAuthorTitle = (LinearLayout) rootView.findViewById(R.id.linear_layout_quote_author_title);
            LinearLayout bookNameTitle = (LinearLayout) rootView.findViewById(R.id.linear_layout_book_name_title);
            LinearLayout publisherNameTitle = (LinearLayout) rootView.findViewById(R.id.linear_layout_publisher_name_title);
            LinearLayout yearNumberTitle = (LinearLayout) rootView.findViewById(R.id.linear_layout_year_number_title);
            LinearLayout pageNumberTitle = (LinearLayout) rootView.findViewById(R.id.linear_layout_page_number_title);
            quoteAuthorTitle.setVisibility(View.GONE);
            bookNameTitle.setVisibility(View.GONE);
            publisherNameTitle.setVisibility(View.GONE);
            yearNumberTitle.setVisibility(View.GONE);
            pageNumberTitle.setVisibility(View.GONE);
        }
    }

    public static CurrentQuoteFragment newInstance(Long currentQuoteTextId, String quoteType) {
        Bundle args = new Bundle();
        args.putSerializable(CURRENT_QUOTE_ID_NEW_INSTANCE, currentQuoteTextId);
        args.putSerializable(CURRENT_QUOTE_TYPE_NEW_INSTANCE, quoteType);
        CurrentQuoteFragment fragment = new CurrentQuoteFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_current_quote, menu);
        setShareAction(menu);
        setDeleteMenuVisibility(menu);
    }

    private void setShareAction(Menu menu) {
        MenuItem item = menu.findItem(R.id.menu_item_share);
        mShareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(item);
        sharingIntent = new Intent(Intent.ACTION_SEND);
        sharingIntent.setType("text/plain");
        String quoteTextForShareBody = mCurrentQuote.first().getQuoteText();
        sharingIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, "It is great quote! Listen!");
        sharingIntent.putExtra(android.content.Intent.EXTRA_TEXT, quoteTextForShareBody);
        if (mShareActionProvider != null) {
            mShareActionProvider.setShareIntent(sharingIntent);
        }
    }

    private void setDeleteMenuVisibility(Menu menu) {
        if (getActivity().findViewById(R.id.detail_fragment_container) != null) {
            MenuItem itemDeleteQuote = menu.findItem(R.id.delete_quote);
            itemDeleteQuote.setVisible(false);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.delete_quote:
                openDeleteQuoteDialog();
                break;
            case R.id.menu_item_share:
                startActivity(Intent.createChooser(sharingIntent, "Select conversation"));
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void openDeleteQuoteDialog() {
        LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
        View dialogView = layoutInflater.inflate(R.layout.dialog_delete, null);
        TextView deleteDialogTitle = (TextView) dialogView.findViewById(R.id.dialog_delete_title);
        deleteDialogTitle.setText(getResources().getString(R.string.dialog_delete_current_quote_title));
        AlertDialog.Builder mDialogBuilder = new AlertDialog.Builder(getActivity());
        mDialogBuilder.setView(dialogView);
        mDialogBuilder
                .setCancelable(false)
                .setPositiveButton(getResources().getString(R.string.dialog_ok_button),
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                mQuoteDataRepository.deleteQuoteByIdFromDb(mCurrentQuoteId, mQuoteType);
                                getActivity().finish();
                            }
                        })
                .setNegativeButton(getResources().getString(R.string.dialog_cancel_button),
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                dialog.cancel();
                            }
                        });
        AlertDialog alertDialog = mDialogBuilder.create();
        alertDialog.show();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(QUOTE_TYPE_BUNDLE, mQuoteType);
        outState.putLong(QUOTE_ID_BUNDLE, mCurrentQuoteId);
    }
}

File: app/src/main/java/com/github/veselinazatchepina/myquotes/currentquote/CurrentQuoteFragment.kt
package com.github.veselinazatchepina.myquotes.currentquote

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.MenuItemCompat
import android.support.v7.widget.ShareActionProvider
import android.view.*
import com.github.veselinazatchepina.myquotes.R
import com.github.veselinazatchepina.myquotes.data.local.entity.BookAuthor
import com.github.veselinazatchepina.myquotes.data.local.entity.BookReleaseYear
import com.github.veselinazatchepina.myquotes.data.local.model.AllQuoteData
import kotlinx.android.synthetic.main.fragment_current_quote.view.*
import java.io.Serializable


class CurrentQuoteFragment : Fragment(), CurrentQuoteContract.View {

    private var currentQuotePresenter: CurrentQuoteContract.Presenter? = null
    var allQuoteData: AllQuoteData? = null
    lateinit var rootView: View
    private var shareIntent: Intent? = null
    var bookAuthors: List<BookAuthor>? = null
    var yearsValues: List<BookReleaseYear>? = null

    companion object {
        private const val ALL_QUOTE_DATA_BUNDLE = "all_quote_data_bundle"

        fun createInstance(quote: AllQuoteData): CurrentQuoteFragment {
            val bundle = Bundle()
            bundle.putSerializable(ALL_QUOTE_DATA_BUNDLE, quote as Serializable)
            val fragment = CurrentQuoteFragment()
            fragment.arguments = bundle
            return fragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        allQuoteData = arguments?.getSerializable(ALL_QUOTE_DATA_BUNDLE) as AllQuoteData
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_current_quote, container, false)
        getBookAuthors()
        getBookReleaseYear()
        showQuote(allQuoteData)
        return rootView
    }

    private fun getBookAuthors() {
        currentQuotePresenter?.getBookAuthors(allQuoteData?.authorsId?.map {
            it.authorIdJoin
        } ?: emptyList())
    }

    private fun getBookReleaseYear() {
        currentQuotePresenter?.getBookReleaseYear(allQuoteData?.yearsId?.map {
            it.yearIdJoin
        } ?: emptyList())
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.menu_current_quote, menu)
        setShareAction(menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    private fun setShareAction(menu: Menu?) {
        val shareItem = menu?.findItem(R.id.menu_item_share)
        val shareActionProvider = MenuItemCompat.getActionProvider(shareItem) as ShareActionProvider?
        shareIntent = Intent(Intent.ACTION_SEND).apply {
            this.type = "text/plain"
            this.putExtra(android.content.Intent.EXTRA_SUBJECT, "It is great quote! Listen!")
            this.putExtra(android.content.Intent.EXTRA_TEXT, allQuoteData?.quote?.quoteText)
        }
        shareActionProvider?.setShareIntent(shareIntent)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.menu_item_share -> {
                startActivity(Intent.createChooser(shareIntent, "Select conversation"))
            }
            R.id.menu_item_delete_quote -> {
                createDeleteQuoteDialog()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun createDeleteQuoteDialog() {
        AlertDialog.Builder(activity).apply {
            val dialogView = LayoutInflater.from(activity).inflate(R.layout.dialog_delete_quote, null)
            this.setView(dialogView)
            this.setCancelable(false)
                    .setPositiveButton(getString(R.string.dialog_add_category_ok)) { dialogInterface: DialogInterface?, id: Int ->
                        currentQuotePresenter?.deleteQuote(allQuoteData?.quote!!.quoteId)
                    }
                    .setNegativeButton(getString(R.string.dialog_add_category_cancel)) { dialogInterface: DialogInterface?, id: Int ->
                        dialogInterface?.cancel()
                    }
        }.create().show()
    }

    override fun setPresenter(presenter: CurrentQuoteContract.Presenter) {
        currentQuotePresenter = presenter
    }

    override fun showBookAuthors(authors: List<BookAuthor>) {
        bookAuthors = authors
        rootView.current_author_name.text = getAllAuthorString(authors)
    }

    override fun showBookReleaseYears(years: List<BookReleaseYear>) {
        yearsValues = years
        rootView.current_year_number.text = isEmptyString(getAllYearString(years))
    }

    private fun showQuote(allQuoteData: AllQuoteData?) {
        rootView.current_quote_text.text = getString(R.string.quote_text_format, allQuoteData?.quote?.quoteText)
        rootView.currentCategory.text = allQuoteData?.category?.first()?.categoryName
        rootView.current_book_name.text = isDataEmpty(allQuoteData?.book!!.bookName, allQuoteData.book!!.bookId, "NoBookName")
        rootView.current_publisher_name.text = isDataEmpty(allQuoteData.publishingOffice!!.first().officeName,
                allQuoteData.publishingOffice!!.first().officeId,
                "NoPublishingOfficeName")
        rootView.current_page_number.text = isEmptyString(allQuoteData.quote?.pageNumber?.toString() ?: "")
        rootView.quote_creation_date.text = isEmptyString(allQuoteData.quote?.creationDate ?: "")
        rootView.comments.text = isEmptyString(allQuoteData.quote?.comments ?: "")
    }

    private fun isEmptyString(currentValue: String): String = if (currentValue == "") {
        "-"
    } else {
        currentValue
    }

    private fun isDataEmpty(currentValue: String, id: Long, pattern: String): String {
        return if (currentValue == pattern + id) {
            "-"
        } else {
            currentValue
        }
    }


    private fun getAllAuthorString(authors: List<BookAuthor>): String {
        var author = ""
        for (i in 0 until authors.size) {
            if (isDataEmpty(authors[i].name, authors[i].authorId, "NoAuthorName") == "-") {
                continue
            }
            author = author.plus("${authors[i].name} " + "${authors[i].surname} " + "${authors[i].patronymic} ")
            if (i != authors.size - 1) {
                author = author.plus(",\n")
            }
        }
        return if (author == "") {
            "-"
        } else {
            author
        }
    }

    private fun getAllYearString(years: List<BookReleaseYear>): String {
        var year = ""
        for (i in 0 until years.size) {
            year = year.plus("${years[i].year}")
            if (i != years.size - 1) {
                year = year.plus(",\n")
            }
        }
        return year
    }
}